src/jit/optimizer.cpp

```diff
@@ -1416,410 +1418,1018 @@ void Compiler::optCheckPreds()
     {
-        top       = head->bbNext;
-        exit      = nullptr;
-        exitCount = 0;
+        // Make sure we've renumbered such that the bitsets can hold all the bits
+        assert(comp->fgBBNumMax <= comp->fgCurBBEpochSize);
+    }
+
+    //------------------------------------------------------------------------
+    // findLoop: Search for a loop with the given HEAD block.
+    //
+    // Arguments:
+    //    head - Block to be the HEAD of any loop identified
+    //
+    // Return Value:
+    //    true  - Found a valid loop.
+    //    false - Did not find a valid loop.
+    //
+    // Notes:
+    //    May modify flow graph go make loop compact before returning.
+    //    Will set instance fields to track loop's extent and exits.
+    //
+    bool findLoop(BasicBlock* head)
+    {
+        BasicBlock* top = head->bbNext;
 
         //  Blocks that are rarely run have a zero bbWeight and should
         //  never be optimized here
 
         if (top->bbWeight == BB_ZERO_WEIGHT)
         {
-            continue;
+            return false;
         }
 
-        for (pred = top->bbPreds; pred; pred = pred->flNext)
+        for (flowList* pred = top->bbPreds; pred; pred = pred->flNext)
         {
-            /* Is this a loop candidate? - We look for "back edges", i.e. an edge from BOTTOM
-             * to TOP (note that this is an abuse of notation since this is not necessarily a back edge
-             * as the definition says, but merely an indication that we have a loop there).
-             * Thus, we have to be very careful and after entry discovery check that it is indeed
-             * the only place we enter the loop (especially for non-reducible flow graphs).
-             */
+            if (findLoop(head, top, pred->flBlock))
+            {
+                return true;
+            }
+        }
+
+        return false;
+    }
+
+    //------------------------------------------------------------------------
+    // recordLoop: Notify the Compiler that a loop has been found.
+    //
+    // Return Value:
+    //    true  - Loop successfully recorded.
+    //    false - Compiler has run out of loop descriptors; loop not recorded.
+    //
+    bool recordLoop()
+    {
+        /* At this point we have a compact loop - record it in the loop table
+        * If we found only one exit, record it in the table too
+        * (otherwise an exit = nullptr in the loop table means multiple exits) */
+
+        BasicBlock* onlyExit = (exitCount == 1 ? lastExit : nullptr);
+        if (comp->optRecordLoop(head, first, top, entry, bottom, onlyExit, exitCount))
+        {
+            // Record the BOTTOM block for future reference before returning.
+            assert(bottom->bbNum < oldBlockMaxNum);
+            BlockSetOps::AddElemD(comp, bottomBlocks, bottom->bbNum);
+            return true;
+        }
+
+        // Unable to record this loop because the loop descriptor table overflowed.
+        return false;
+    }
```

Comment:
If the loop table is full, wouldn't we know that before we did all the work to detect the loop? I.e., why do all that work if we're just going to fail to add it to the loop table?